Incorrect LIBPATH for address-model=32,64 on Windows using libjpeg.jam libpng.jam libtiff.jam zlib.jam

I'm building Boost.GIL tests on Windows which depend on third-parties like libjpeg, libpng, etc.

I've configured `user-config.jam` to consume those dependencies installed by vcpkg package manager.

When requesting multiple architectures using `address-model=32,64` or `address-model=64,32`, targets for the second one fail to link due to incorrect `LIBPATH` in B2 command line.
IOW, requesting single architecture  with `b2.exe address-model=64` or `b2.exe address-model=32` works fine.

Initially, I posted the problem to https://lists.boost.org/boost-build/2019/12/30275.php, but following brief investigation on Slack with @pdimov, there seem to be a problem in B2.

### `%HOME%\user-config.jam`

```
project
  : requirements
    <address-model>32:<dll-path>C:/vcpkg/installed/x86-windows/bin
    <address-model>64:<dll-path>C:/vcpkg/installed/x64-windows/bin
  ;

using libjpeg
  :
  : <include>C:/vcpkg/installed/x86-windows/include <search>C:/vcpkg/installed/x86-windows/lib
  : <address-model>32
  ;

using libjpeg
  :
  : <include>C:/vcpkg/installed/x64-windows/include <search>C:/vcpkg/installed/x64-windows/lib
  : <address-model>64
  ;

using libpng
  :
  : <include>C:/vcpkg/installed/x86-windows/include <search>C:/vcpkg/installed/x86-windows/lib
  : <address-model>32
  ;

using libpng
  :
  : <include>C:/vcpkg/installed/x64-windows/include <search>C:/vcpkg/installed/x64-windows/lib
  : <address-model>64
  ;

using libtiff
  :
  : <include>C:/vcpkg/installed/x86-windows/include <search>C:/vcpkg/installed/x86-windows/lib
  : <address-model>32
  ;

using libtiff
  :
  : <include>C:/vcpkg/installed/x64-windows/include <search>C:/vcpkg/installed/x64-windows/lib
  : <address-model>64
  ;

using zlib
  :
  : <include>C:/vcpkg/installed/x86-windows/include <search>C:/vcpkg/installed/x86-windows/lib
  : <address-model>32
  ;

using zlib
  :
  : <include>C:/vcpkg/installed/x64-windows/include <search>C:/vcpkg/installed/x64-windows/lib
  : <address-model>64
  ;
```

### Problem 1: `address-model=32,64` linking 64-bit target against 32-bit libraries

    ```console
    cd D:\boost.win
    D:\boost.win> rmdir /s /q bin.v2
    D:\boost.win> b2.exe toolset=msvc-14.2 address-model=32,64 libs/gil/test/extension/io//simple

    Performing configuration checks

        - default address-model    : 32-bit
        - default architecture     : x86
        - symlinks supported       : no
        - junctions supported      : yes
        - hardlinks supported      : yes
        - Boost.Config Feature Check: cxx11_constexpr : yes
        - Boost.Config Feature Check: cxx11_defaulted_functions : yes
        - Boost.Config Feature Check: cxx11_template_aliases : yes
        - Boost.Config Feature Check: cxx11_trailing_result_types : yes
        - Boost.Config Feature Check: cxx11_variadic_templates : yes
        - Boost.Config Feature Check: cxx11_constexpr : yes
        - Boost.Config Feature Check: cxx11_defaulted_functions : yes
        - Boost.Config Feature Check: cxx11_template_aliases : yes
        - Boost.Config Feature Check: cxx11_trailing_result_types : yes
        - Boost.Config Feature Check: cxx11_variadic_templates : yes
        - BOOST_COMP_GNUC >= 4.3.0 : no
        - libjpeg                  : yes
        - zlib                     : yes
        - libpng                   : yes
        - libtiff                  : yes
        - Boost.Config Feature Check: cxx11_constexpr : yes
        - Boost.Config Feature Check: cxx11_defaulted_functions : yes
        - Boost.Config Feature Check: cxx11_template_aliases : yes
        - Boost.Config Feature Check: cxx11_trailing_result_types : yes
        - Boost.Config Feature Check: cxx11_variadic_templates : yes
        - Boost.Config Feature Check: cxx11_constexpr : yes
        - Boost.Config Feature Check: cxx11_defaulted_functions : yes
        - Boost.Config Feature Check: cxx11_template_aliases : yes
        - Boost.Config Feature Check: cxx11_trailing_result_types : yes
        - Boost.Config Feature Check: cxx11_variadic_templates : yes
        - BOOST_COMP_GNUC >= 4.3.0 : no
        - libjpeg                  : yes
        - zlib                     : yes
        - libpng                   : yes
        - libtiff                  : yes
    ...
    C:\vcpkg\installed\x86-windows\lib\jpeg.lib : warning LNK4272: library machine type 'x86' conflicts with target machine type 'x64'
    C:\vcpkg\installed\x86-windows\lib\zlib.lib : warning LNK4272: library machine type 'x86' conflicts with target machine type 'x64'
    C:\vcpkg\installed\x86-windows\lib\libpng16.lib : warning LNK4272: library machine type 'x86' conflicts with target machine type 'x64'
    C:\vcpkg\installed\x86-windows\lib\tiff.lib : warning LNK4272: library machine type 'x86' conflicts with target machine type 'x64'
    bin.v2\libs\gil\test\extension\io\all_formats_test.test\msvc-14.2\debug\address-model-64\asynch-exceptions-on\threading-multi\all_formats_test.exe : fatal error LNK1120: 107 unresolved externals

            call "bin.v2\standalone\msvc\msvc-14.2\address-model-64\architecture-x86\msvc-setup.bat"  >nul
    link /NOLOGO /INCREMENTAL:NO /DEBUG /MACHINE:X64 /MANIFEST /subsystem:console 
        /out:"bin.v2\libs\gil\test\extension\io\all_formats_test.test\msvc-14.2\debug\address-model-64\asynch-exceptions-on\threading-multi\all_formats_test.exe"
        /LIBPATH:"C:\vcpkg\installed\x86-windows\lib"  
        @"bin.v2\libs\gil\test\extension\io\all_formats_test.test\msvc-14.2\debug\address-model-64\asynch-exceptions-on\threading-multi\all_formats_test.exe.rsp"
    ```

### Problem 2:  `address-model=64,32` linking 32-bit target against 64-bit libraries

    ```console
    cd D:\boost.win
    rmdir /s /q bin.v2
    b2.exe toolset=msvc-14.2 address-model=64,32 libs/gil/test/extension/io//simple
    Performing configuration checks

        - default address-model    : 32-bit
        - default architecture     : x86
        - symlinks supported       : no
        - junctions supported      : yes
        - hardlinks supported      : yes
        - Boost.Config Feature Check: cxx11_constexpr : yes
        - Boost.Config Feature Check: cxx11_defaulted_functions : yes
        - Boost.Config Feature Check: cxx11_template_aliases : yes
        - Boost.Config Feature Check: cxx11_trailing_result_types : yes
        - Boost.Config Feature Check: cxx11_variadic_templates : yes
        - Boost.Config Feature Check: cxx11_constexpr : yes
        - Boost.Config Feature Check: cxx11_defaulted_functions : yes
        - Boost.Config Feature Check: cxx11_template_aliases : yes
        - Boost.Config Feature Check: cxx11_trailing_result_types : yes
        - Boost.Config Feature Check: cxx11_variadic_templates : yes
        - BOOST_COMP_GNUC >= 4.3.0 : no
        - libjpeg                  : yes
        - zlib                     : yes
        - libpng                   : yes
        - libtiff                  : yes
        - Boost.Config Feature Check: cxx11_constexpr : yes
        - Boost.Config Feature Check: cxx11_defaulted_functions : yes
        - Boost.Config Feature Check: cxx11_template_aliases : yes
        - Boost.Config Feature Check: cxx11_trailing_result_types : yes
        - Boost.Config Feature Check: cxx11_variadic_templates : yes
        - Boost.Config Feature Check: cxx11_constexpr : yes
        - Boost.Config Feature Check: cxx11_defaulted_functions : yes
        - Boost.Config Feature Check: cxx11_template_aliases : yes
        - Boost.Config Feature Check: cxx11_trailing_result_types : yes
        - Boost.Config Feature Check: cxx11_variadic_templates : yes
        - BOOST_COMP_GNUC >= 4.3.0 : no
        - libjpeg                  : yes
        - zlib                     : yes
        - libpng                   : yes
        - libtiff                  : yes
    ...
    C:\vcpkg\installed\x64-windows\lib\jpeg.lib : warning LNK4272: library machine type 'x64' conflicts with target machine type 'x86'
    C:\vcpkg\installed\x64-windows\lib\zlib.lib : warning LNK4272: library machine type 'x64' conflicts with target machine type 'x86'
    C:\vcpkg\installed\x64-windows\lib\libpng16.lib : warning LNK4272: library machine type 'x64' conflicts with target machine type 'x86'
    C:\vcpkg\installed\x64-windows\lib\tiff.lib : warning LNK4272: library machine type 'x64' conflicts with target machine type 'x86'
    bin.v2\libs\gil\test\extension\io\all_formats_test.test\msvc-14.2\debug\asynch-exceptions-on\threading-multi\all_formats_test.exe : fatal error LNK1120: 107 unresolved externals

            call "bin.v2\standalone\msvc\msvc-14.2\msvc-setup.bat"  >nul
    link /NOLOGO /INCREMENTAL:NO /DEBUG /MACHINE:X86 /MANIFEST /subsystem:console 
         /out:"bin.v2\libs\gil\test\extension\io\all_formats_test.test\msvc-14.2\debug\asynch-exceptions-on\threading-multi\all_formats_test.exe"
         /LIBPATH:"C:\vcpkg\installed\x64-windows\lib"  
         @"bin.v2\libs\gil\test\extension\io\all_formats_test.test\msvc-14.2\debug\asynch-exceptions-on\threading-multi\all_formats_test.exe.rsp"
    ```

### Boost.GIL Jamfile-s

A word on the Jamfiles that are used above, may be relevant:

GIL's [test/extension/io/Jamfile](https://github.com/boostorg/gil/blob/9e76b8b5bf7a16c50b7940c969e473855098f337/test/extension/io/Jamfile#L13-L16) has these `using`-s which, as I have been told, should not really be in `Jamfile`:

```
using libjpeg : : : : true ; # work around bug on master
using zlib ;
using libpng : : : : true ;
using libtiff : : : : true ;
```

I removed them and tried the commands above. No difference, so I assume presence of those is benign.

--- a/src/tools/generators/searched-lib-generator.jam
+++ b/src/tools/generators/searched-lib-generator.jam
@@ -38,7 +38,7 @@ class searched-lib-generator : generator
 
             local search = [ feature.get-values <search> : $(properties) ] ;
 
-            local a = [ new null-action [ $(property-set).add-raw <relevant>link ] ] ;
+            local a = [ new null-action [ $(property-set).add-raw <relevant>link <relevant>search ] ] ;
             local lib-name = [ feature.get-values <name> : $(properties) ] ;
             lib-name ?= $(name) ;
             local t = [ new searched-lib-target $(lib-name) : $(project)
@@ -47,7 +47,7 @@ class searched-lib-generator : generator
             #    lib png : z : <name>png ;
             # the 'z' target should be returned, so that apps linking to 'png'
             # will link to 'z', too.
-            return [ property-set.create <xdll-path>$(search) <relevant>link ]
+            return [ property-set.create <xdll-path>$(search) <relevant>link <relevant>search <relevant>name ]
                    [ virtual-target.register $(t) ] $(sources) ;
         }
     }
@@ -88,7 +88,7 @@ class searched-lib-target : abstract-file-target
 
     rule relevant ( )
     {
-        return [ property-set.create <relevant>link ] ;
+        return [ property-set.create <relevant>link <relevant>search ] ;
     }
 
     rule path ( )

--- a/test/searched_lib.py
+++ b/test/searched_lib.py
@@ -163,6 +163,22 @@ lib l : : <name>l_f ;
 t.run_build_system(["-n"])
 
 
+# A regression test. Virtual targets distinguished by their search paths were
+# not differentiated when registered, which caused search paths to be selected
+# incorrectly for build requests with multiple feature values.
+t.write("jamroot.jam", "")
+t.write("a.cpp", "")
+t.write("jamfile.jam", """\
+exe a : a.cpp l ;
+lib l : : <name>l <search>lib32 <address-model>32 ;
+lib l : : <name>l <search>lib64 <address-model>64 ;
+""")
+
+t.run_build_system(["-n","address-model=32,64"])
+t.fail_test(t.stdout().find("lib32") == -1)
+t.fail_test(t.stdout().find("lib64") == -1)
+
+
 # Make sure plain "lib foobar ; " works.
 t.write("jamfile.jam", """\
 exe a : a.cpp foobar ;

